staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go

```diff
@@ -138,5 +151,16 @@ func (e *streamExecutor) Stream(options StreamOptions) error {
 		streamer = newStreamProtocolV1(options)
 	}
 
-	return streamer.stream(conn)
+	errorChan := make(chan error, 1)
+	go func() {
+		defer close(errorChan)
+		errorChan <- streamer.stream(conn)
```

Comment:
does canceling the context close the conn/stream operation or does it leave this goroutine hung?

Reply:
can streamer.stream panic? if so, spawning a goroutine here means the process will exit if that happens, as opposed to propagating to any panic recovery code put in place by the `Stream` caller

Reply:
> does canceling the context close the conn/stream operation or does it leave this goroutine hung?

Yes, the stream closes after the connection closed (in the deferred statement above).

Reply:
> can streamer.stream panic? if so, spawning a goroutine here means the process will exit if that happens, as opposed to propagating to any panic recovery code put in place by the `Stream` caller

No I don't think so. But still we could add a panic handler.

Reply:
@liggitt It's not possible to propagate the panic to caller since the `stream` goroutine may be still running when the method returns at `return ctx.Err()`. By the way, they did use `runtime.HandleCrash` in goroutines such what in `streamProtocolV2.copyStdout` as well, which is started by `stream`.

```go
func (p *streamProtocolV2) copyStdout(wg *sync.WaitGroup) {
	if p.Stdout == nil {
		return
	}

	wg.Add(1)
	go func() {
		defer runtime.HandleCrash()
		defer wg.Done()
		// make sure, packet in queue can be consumed.
		// block in queue may lead to deadlock in conn.server
		// issue: https://github.com/kubernetes/kubernetes/issues/96339
		defer io.Copy(ioutil.Discard, p.remoteStdout)

		if _, err := io.Copy(p.Stdout, p.remoteStdout); err != nil {
			runtime.HandleError(err)
		}
	}()
}
```

So if this panics, the process exits as well. AFAIK, there's no unexpected panic unless I've missed a bug (and it hasn't been triggered so far). So instead of being defensive, I think it's good to let it just panic and quit so that we can locate the bug immediately and have a quick fix. Logging is the best that we can do.

Reply:
> @liggitt It's not possible to propagate the panic to caller since the stream goroutine may be still running when the method returns at return ctx.Err().

it *is* possible, like this:

```diff
diff --git a/staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go b/staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go
index 4e22a970c35..da6a267b83d 100644
--- a/staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go
+++ b/staging/src/k8s.io/client-go/tools/remotecommand/remotecommand.go
@@ -22,6 +22,7 @@ import (
 	"io"
 	"net/http"
 	"net/url"
+	"sync"
 
 	"k8s.io/klog/v2"
 
@@ -155,6 +156,15 @@ func (e *streamExecutor) newConnectionAndStream(ctx context.Context, options Str
 // StreamWithContext opens a protocol streamer to the server and streams until a client closes
 // the connection or the server disconnects or the context is done.
 func (e *streamExecutor) StreamWithContext(ctx context.Context, options StreamOptions) error {
+	var caughtPanicLock sync.Mutex
+	var caughtPanic error
+	defer func() {
+		caughtPanicLock.Lock()
+		defer caughtPanicLock.Unlock()
+		if caughtPanic != nil {
+			panic(caughtPanic)
+		}
+	}()
 	conn, streamer, err := e.newConnectionAndStream(ctx, options)
 	if err != nil {
 		return err
@@ -163,6 +173,21 @@ func (e *streamExecutor) StreamWithContext(ctx context.Context, options StreamOp
 
 	errorChan := make(chan error, 1)
 	go func() {
+		defer func() {
+			err := recover()
+			if err == nil {
+				return
+			}
+			// Same as stdlib http server code. Manually allocate stack
+			// trace buffer size to prevent excessively large logs
+			const size = 64 << 10
+			buf := make([]byte, size)
+			buf = buf[:runtime.Stack(buf, false)]
+
+			caughtPanicLock.Lock()
+			defer caughtPanicLock.Unlock()
+			caughtPanic = fmt.Errorf("%v\n%s", err, buf)
+		}()
 		defer runtime.HandleCrash()
 		defer close(errorChan)
 		errorChan <- streamer.stream(conn)
```

That ensures existing callers (using context.Background()) have panics propagated as before instead of exiting the process, right?

Reply:
Yes, I agree that we can propagate panics like this. But it doesn't handle the cases when contexts can be canceled, in which the lock can't block the return of `StreamWithContext` and the panic might happen after that. What about using a channel? Like the following:

```diff
@@ -155,6 +155,13 @@ func (e *streamExecutor) newConnectionAndStream(ctx context.Context, options Str
 // StreamWithContext opens a protocol streamer to the server and streams until a client closes
 // the connection or the server disconnects or the context is done.
 func (e *streamExecutor) StreamWithContext(ctx context.Context, options StreamOptions) error {
+       panicChan := make(chan any, 1)
+       defer func() {
+               if r := <-panicChan; r != nil {
+                       panic(r)
+               }
+       }()
+
        conn, streamer, err := e.newConnectionAndStream(ctx, options)
        if err != nil {
                return err
@@ -163,6 +170,12 @@ func (e *streamExecutor) StreamWithContext(ctx context.Context, options StreamOp

        errorChan := make(chan error, 1)
        go func() {
+               defer func() {
+                       if r := recover(); r != nil {
+                               panicChan <- r
+                       }
+                       close(panicChan)
+               }()
                defer runtime.HandleCrash()
                defer close(errorChan)
                errorChan <- streamer.stream(conn)
```

Do we only care about callers who pass a background context, or do we want to cover all the cases? If it's the first one, we can have a separate branch that runs the `streamer.stream` directly when `Context` is not cancellable, which can be determined with `Context.Done() == nil`. If not and we still want to propagate the panic, then I think blocking the caller with a channel in defer is a feasible solution. But still, I have to say, we can not propagate panics inside the goroutines started by `streamer.stream` unless we modify them in such a way as well.

What do you think?

Reply:
>  in which the lock can't block the return of StreamWithContext and the panic might happen after that.

at that time this has already returned, do we want to crash or swallow the panic? it seems that trying to cover everything sounds too much ... and this scenario seems very rare

Reply:
> > in which the lock can't block the return of StreamWithContext and the panic might happen after that.
> 
> at that time this has already returned, do we want to crash or swallow the panic? it seems that trying to cover everything sounds too much ... and this scenario seems very rare

IMO, I think it's ok not to handle the panic because I believe it won't, as far as I understand. And I agree that we don't have to cover everything, but I prefer to keep the behavior predictable. So my suggestion would be either we use `runtime.HandleCrash` here as others did and do nothing more, or we propagate the panic here using a channel. WDYT?

Reply:
I mean, we can propagate the panic, I think that is what Jordan suggested and is ok, but as you said there is still a case where the panic could not be propagated because we already returned, my question is about that case, what are we going to do it, crash the program or ignore it because the context was already cancelled?

Reply:
basically I'm asking if jordan's suggestion with the var and mutex is enough ... this is becoming more complex, but Jordan intuition use to be right

Reply:
I agree it's complex, I just don't have visibility to how this library is being used out of tree, and what the increased risk of unhandled panic is, and what the implications for those consumers are :-/

Reply:
nah, panic should be handled for sure, if it is used in a handler we can not shutdown the whole server .. does it means that we can swallow the panic after the context has cancelled 🤔 ?

Reply:
Unless we have something like the structured concurrency in Go, we always increase the risk of unhandled panic when starting a new goroutine. 

I agree that to make sure the program does not exit accidentally, the best choice is to handle the panic thrown by the new goroutine. But I don't think we can swallow the panic. It could confuse the programmer who uses this method:

```go
ctx, cancel := context.WithTimeout(context.Background(), 5 * time.Second)
defer cancel()
// If the panic happens after the timeout, then err = context.ErrDeadlineExceeded and 
// the panic will be swallowed. 
//
// But if the panic happens before the timeout, then the panic is propagated to the caller.
err := exec.StreamWithContext(ctx, opts)
```

From the users' perspective, they will think these are random panics, not they have gotten something wrong. Therefore, I would prefer to propagate the panic in any case. And we can keep the behavior consistent before and after this change. 🤔 

If this makes sense to you, I'm happy to add the codes.

Reply:
sgmt, wait for @liggitt ack

Reply:
@liggitt I just found we have this pattern https://github.com/kubernetes/kubernetes/pull/103177/files#discussion_r1001353834 in 

https://github.com/kubernetes/kubernetes/blob/4a3bac5eaea81cef0c7b1dcbf47c9eb5c58a30af/staging/src/k8s.io/client-go/tools/cache/reflector.go#L368-L420

WDYT about using the channel instead of the lock?

Reply:
I don't feel strongly, whatever is clearer

Reply:
I see. So let me propagate the panic with a channel, shall I? @liggitt @aojea

Reply:
channel is used in the std library https://github.com/golang/go/blob/81efd7b347dd6d7f12fd49c6eee0274005734c71/src/net/http/server.go#L3356-L3383

and in other projects too https://grep.app/search?q=panicCh&filter[lang][0]=Go

Reply:
let's go with channel

Reply:
Done! Please take a look. @aojea @liggitt 

I followed the pattern from https://github.com/golang/go/blob/81efd7b347dd6d7f12fd49c6eee0274005734c71/src/net/http/server.go#L3356-L3383